tracer/src/Datadog.Trace/Iast/ReturnedHeadersAnalyzer.cs

```diff
@@ -39,6 +41,140 @@ internal static void Analyze(IHeaderDictionary responseHeaders, IntegrationId in
         AnalyzeXContentTypeOptionsVulnerability(responseHeaders, integrationId, serviceName, responseCode);
         AnalyzeStrictTransportSecurity(responseHeaders, integrationId, serviceName, responseCode, protocol);
         AnalyzeUnvalidatedRedirect(responseHeaders, integrationId);
+        AnalyzeHeaderInjectionVulnerability(responseHeaders, integrationId);
+    }
+
+    // In header injections, we should exclude some headers to prevent false positives:
+    // location: it is already reported in UNVALIDATED_REDIRECT vulnerability detection.
+    // Sec-WebSocket-Location, Sec-WebSocket-Accept, Upgrade, Connection: Usually the framework gets info from request
+    // access-control-allow-*: when the source of the tainted range is the request header origin or access-control-request-headers
+    // set-cookie: We should ignore set-cookie header if the source of all the tainted ranges are cookies
+    // We should exclude the injection when the tainted string only has one range which comes from a request header with the same name that the header that we are checking in the response.
+    // Headers could store sensitive information, we should redact whole <header_value> if:
+    // <header_name> matches with a RegExp
+    // <header_value> matches with  a RegExp
+    // We should redact the sensitive information from the evidence when:
+    // Tainted range is considered sensitive value
+
+#if NETFRAMEWORK
+    private static void AnalyzeHeaderInjectionVulnerability(NameValueCollection responseHeaders, IntegrationId integrationId)
+#else
+    private static void AnalyzeHeaderInjectionVulnerability(IHeaderDictionary responseHeaders, IntegrationId integrationId)
+#endif
+    {
+        try
+        {
+            IastModule.OnExecutedSinkTelemetry(IastInstrumentedSinks.HeaderInjection);
+
+            var tracer = Tracer.Instance;
+            var currentSpan = (tracer.ActiveScope as Scope)?.Span;
+            var traceContext = currentSpan?.Context?.TraceContext;
+
+#if NETFRAMEWORK
+            foreach (var headerKey in responseHeaders.AllKeys)
+#else
+            foreach (var headerKey in responseHeaders.Keys)
+#endif
+            {
+                if (string.IsNullOrWhiteSpace(headerKey))
+                {
+                    continue;
+                }
+
+                bool isHeaderInjectionException = false;
+                foreach (var excludeType in headerInjectionExceptions)
+                {
+                    if (excludeType.Equals((string)headerKey, StringComparison.OrdinalIgnoreCase))
+                    {
+                        isHeaderInjectionException = true;
+                        break;
+                    }
+                }
+
+                if (isHeaderInjectionException)
+                {
+                    continue;
+                }
+
+                string headerValue = responseHeaders[headerKey];
```

Comment:
`IHeaderDictionary` is a dictionary, so it would be better to enumerate the collection directly instead of using `Keys`. Then you don't need this second access.

That would require you extract a helper method to handle the different approach for .NET FX and .NET Core

Reply:
I did this for simplicity but it's true, this access can be avoided. I have updated this method to avoid this access.